source/server/worker_impl.cc

```diff
@@ -135,6 +135,7 @@ void WorkerImpl::threadRoutine(GuardDog& guard_dog) {
   dispatcher_->run(Event::Dispatcher::RunType::Block);
   ENVOY_LOG(debug, "worker exited dispatch loop");
   guard_dog.stopWatching(watch_dog_);
+  dispatcher_->shutdown();
```

Comment:
It seems to me that the only dispatcher that needs shutdown called on it based on the changes in this PR is the dispatcher associated with the main thread.

Reply:
For bug fix, yes, only main thread dispatcher is needed.

Add shutdown here to detect future complicated object destroy. Ideally, I want to add more states in shutdown() and verify it in ~Dispatcher.

Reply:
You're getting to what the long term roadmap should be.  Requiring shutdown be called before dispatcher deletion seems like an undue burden in most cases, specially tests.

Reply:
No, I don't require shutdown() to be called everywhere. The TODO in `~DispatcherImpl()` requires `shutdown()` called. 

Calling `shutdown()` in the worker thread is not a brutal move. `shutdown()` does not execute strict check. Just delete drain the list and flip the flag. All tests are passing right now, aren't they? :)

Reply:
Stepping back can this just be done by the destructor? Do we need a shutdown method?

Reply:
The deletion of the dispatcher would have to happen in the places where shutdown is called in this PR.  The early deletion of the dispatcher would result in use-after-free as other threads attempt to post to the deleted dispatcher.

So yes, I think we need a shutdown method.

Reply:
OK sounds good. I don't have a strong opinion on whether to do it only on the main thread vs. everywhere. Either way seems fine to me.

Reply:
We need shutdown() prior to ~Dispatcher in master thread as @antoniovicente explained. 

For worker thread at this moment, we don't need shutdown(). But a shutdown() plus the future assertion in `~Dispatcher` may prevent regression.